File: python_practice/48.旋转图像.py
#
# @lc app=leetcode.cn id=48 lang=python3
#
# [48] 旋转图像
#
from typing import List
import math

# @lc code=start
class Solution:
    def rotate(self, matrix: List[List[int]]) -> None:
        """
        Do not return anything, modify matrix in-place instead.
        """
        n = len(matrix)
        round = math.ceil(n // 2)
        # rounds of rotation
        for i in range(round):
            # movement of elements in one round
            for j in range(i, n-1-i):
                temp = matrix[i][j]
                # left up <- left down 
                matrix[i][j] = matrix[n-1-j][i]
                # left down <- right down
                matrix[n-1-j][i] = matrix[n-1-i][n-1-j]
                # right down <- right up
                matrix[n-1-i][n-1-j] = matrix[j][n-1-i]
                # right up <- left up
                matrix[j][n-1-i] = temp
 
# @lc code=end

File: cpp/48.旋转图像.cpp
/*
 * @lc app=leetcode.cn id=48 lang=cpp
 *
 * [48] 旋转图像
 */

// @lc code=start
class Solution {
public:
    // 顺时针90°：转置 + 水平镜像
    // 逆时针90°：转置 + 垂直镜像
    // 180°：水平镜像 + 垂直镜像
    void rotate(vector<vector<int>>& matrix) {
        for (int i = 0; i < matrix.size(); ++i) {
            // 注意这里j < i
            for (int j = 0; j < i; ++j) {
                swap(matrix[i][j], matrix[j][i]);
            }
        }

        for (int i = 0; i < matrix.size(); ++i) {
            for (int j = 0, k = matrix.size() - 1; j < k; ++j, --k) {
                swap(matrix[i][j], matrix[i][k]);
            }
        }
        
    }
};
// @lc code=end

